src/audio/drc/drc.c

```diff
@@ -0,0 +1,606 @@
+
+	if (state->last_pre_delay_frames != pre_delay_frames) {
+		state->last_pre_delay_frames = pre_delay_frames;
+		state->pre_delay_read_index = 0;
+		state->pre_delay_write_index = pre_delay_frames;
+	}
+	return 0;
+}
+
+static int drc_setup(struct comp_data *cd, uint16_t channels, uint32_t rate)
+{
+	int ret = 0;
+
+	/* Reset any previous state */
+	drc_reset_state(&cd->state);
+
+	/* Allocate pre-delay buffers */
+	uint32_t sample_bytes = get_sample_bytes(cd->source_format);
+	ret = drc_init_pre_delay_buffers(&cd->state, (size_t)sample_bytes, (int)channels);
+	if (ret < 0)
+		return ret;
+
+	/* Set pre-dely time */
+	return drc_set_pre_delay_time(&cd->state, cd->config->params.pre_delay_time, rate);
+}
+
+/*
+ * End of DRC setup code. Next the standard component methods.
+ */
+
+static struct comp_dev *drc_new(const struct comp_driver *drv,
+				struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct comp_data *cd;
+	struct sof_ipc_comp_process *drc;
+	struct sof_ipc_comp_process *ipc_drc =
+		(struct sof_ipc_comp_process *)comp;
+	size_t bs = ipc_drc->size;
+	int ret;
+
+	comp_cl_info(&comp_drc, "drc_new()");
+
+	/* Check first before proceeding with dev and cd that coefficients
+	 * blob size is sane.
+	 */
+	if (bs > SOF_DRC_MAX_SIZE) {
+		comp_cl_err(&comp_drc, "drc_new(), coefficients blob size %u exceeds maximum",
+			    bs);
+		return NULL;
+	}
+
+	dev = comp_alloc(drv, COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+
+	drc = COMP_GET_IPC(dev, sof_ipc_comp_process);
+	ret = memcpy_s(drc, sizeof(*drc), ipc_drc,
+		       sizeof(struct sof_ipc_comp_process));
+	assert(!ret);
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	cd->drc_func = NULL;
+	cd->config = NULL;
+	cd->config_new = NULL;
+
+	if (bs) {
+		cd->config = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				     bs);
+		if (!cd->config) {
+			rfree(dev);
+			rfree(cd);
+			return NULL;
+		}
+
+		ret = memcpy_s(cd->config, bs, ipc_drc->data, bs);
+		assert(!ret);
+	}
+
+	drc_reset_state(&cd->state);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void drc_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "drc_free()");
+
+	drc_free_parameters(&cd->config);
+	drc_free_parameters(&cd->config_new);
+
+	rfree(cd);
+	rfree(dev);
+}
+
+static int drc_verify_params(struct comp_dev *dev,
+			     struct sof_ipc_stream_params *params)
+{
+	int ret;
+
+	comp_dbg(dev, "drc_verify_params()");
+
+	ret = comp_verify_params(dev, 0, params);
+	if (ret < 0) {
+		comp_err(dev, "drc_verify_params(): comp_verify_params() failed.");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int drc_params(struct comp_dev *dev,
+		      struct sof_ipc_stream_params *params)
+{
+	int err;
+
+	comp_info(dev, "drc_params()");
+
+	err = drc_verify_params(dev, params);
```

Comment:
I don't think `drc_verify_params` needs to be its own function, all it does is call another function